### PR TITLE
feat: validate blocks against replay set

### PR DESCRIPTION
**Pull request by hstove**
- Closes https://github.com/stacks-network/stacks-core/issues/6088

This PR implements:

- Validating that a proposed block contains the replay set (if we're in replay mode)
- Clearing the replay set if we're "done" with transaction replay
- Adds a feature flag to the signer for whether or not to validate blocks against the replay set. We'll want to remove this feature flag as part of https://github.com/stacks-network/stacks-core/issues/5965?issue=stacks-network%7Cstacks-core%7C6087

Knowing if we should be "done" with replay is tricky. What this implementation does is add a field to `BlockValidateOk` with a hash of the replay set, along with whether this block exhausts that replay set. This field is only present if the block was validated against a replay set. Basically, it's saying "after this block, we can move out of replay". The signer stores this (block_hash, replay_set, replay_set_exhausted) tuple. Then, in `stacks_block_arrival`, the signer state machine checks if we should move out of replay based on this lookup.

I've added a test (`tx_replay_e2e_test`) which is mainly @jferrant's code from https://github.com/stacks-network/stacks-core/pull/6031/files#diff-9fbead3ef50a4b19fa3c1ce9c11952010f8506706db040df2d85459fe8d518b7 demonstrating:

- bitcoin forks, we go into replay mode
- miner tries to make a non-replay block, gets rejected
- miner makes a replay block, gets approved
- replay set is cleared after
- regular mining resumes

This PR is definitely missing some edge cases, and likely will break other tests, but I wanted to get some eyes on it first.

### CODE REVIEW

**Review by fdefelici — COMMENTED** [review]
Completed a first review round: added a bunch of nits and some checks that could lead to code improvements

File: stacks-signer/src/signerdb.rs [hunk 1]
```diff
@@ -708,6 +715,11 @@ static SCHEMA_14: &[&str] = &[
...
 ];
 
+static SCHEMA_15: &[&str] = &[
+    ADD_BLOCK_VALIDATED_BY_REPLAY_TXS_TABLE,
+    "INSERT INTO db_config (version) VALUES (15);",
+];
+
 impl SignerDb {
     /// The current schema version used in this build of the signer binary.
     pub const SCHEMA_VERSION: u32 = 12;
```
[the remark this answers is not in the document]
Comment (kantai) [reply stacks-signer/src/signerdb.rs:725]: oops!

[the remark this answers is not in the document]
Comment (kantai) [reply stacks-signer/src/signerdb.rs:725]: I think this needs to be fixed in `develop` too

[the remark this answers is not in the document]
Comment (kantai) [reply stacks-signer/src/signerdb.rs:725]: (Actually, it doesn't **really** need to be, because its only used in an error message)

File: stackslib/src/net/api/postblock_proposal.rs [hunk 2]
```diff
@@ -734,6 +755,19 @@ impl NakamotoBlockProposal {
...
+        })
+    }
+
+    pub fn tx_replay_hash(&self) -> Option<u64> {
+        self.replay_txs.as_ref().map(|txs| {
+            let mut hasher = DefaultHasher::new();
+            txs.hash(&mut hasher);
+            hasher.finish()
         })
     }
```
Comment (kantai) [inline stackslib/src/net/api/postblock_proposal.rs:772]: This should be reused between the signer and the stacks-node binaries.

File: stacks-signer/src/v0/signer_state.rs [hunk 3]
```diff
@@ -366,10 +369,22 @@ impl LocalStateMachine {
...
 
-        // No matter what, if we're in tx replay mode, remove the tx replay set
-        // TODO: in later versions, we will only clear the tx replay
-        // set when replay is completed.
-        prior_state_machine.tx_replay_set = None;
+        if let Some(tx_replay_set) = prior_state_machine.tx_replay_set.clone() {
+            // let replay_set_hash = hash_tx_replay_set(tx_replay_set);
+            let mut hasher = DefaultHasher::new();
+            tx_replay_set.hash(&mut hasher);
+            let replay_set_hash = hasher.finish();
```
Comment (kantai) [inline stacks-signer/src/v0/signer_state.rs:376]: Can this invoke the `postblock_proposal::tx_replay_hash` function?

File: stacks-signer/src/v0/signer_state.rs [hunk 4]
```diff
@@ -366,10 +369,22 @@ impl LocalStateMachine {
...
+            if let Ok(true) =
+                db.get_was_block_validated_by_replay_tx(signer_signature_hash, replay_set_hash)
```
Comment (kantai) [inline stacks-signer/src/v0/signer_state.rs:378]: Hmm, this `if let` kind of scares me even though I know it works. `if let Ok(t) = ..`, for example, would compile, but do something different.

I'd prefer just doing:

```suggestion
            if Ok(true) ==
                db.get_was_block_validated_by_replay_tx(signer_signature_hash, replay_set_hash)
```

File: stacks-signer/src/v0/signer_state.rs [hunk 5]
```diff
@@ -366,10 +369,22 @@ impl LocalStateMachine {
             }
         };
 
-        // No matter what, if we're in tx replay mode, remove the tx replay set
-        // TODO: in later versions, we will only clear the tx replay
-        // set when replay is completed.
-        prior_state_machine.tx_replay_set = None;
+        if let Some(tx_replay_set) = prior_state_machine.tx_replay_set.clone() {
```
Comment (kantai) [inline stacks-signer/src/v0/signer_state.rs:372]: Can the clone be avoided?

File: stacks-signer/src/v0/signer_state.rs (hunk 4 above)
Comment (kantai) [inline stacks-signer/src/v0/signer_state.rs:378]: What happens to signers who don't validate block A (maybe they were offline?), where block A is the block that exhausts the replay set. If they never _process_ A, then the question doesn't really matter, because they're going to be behind chain tip anyways. But if they catch up to chain tip, they'll have processed A, but still think the replay set is active, right?

**Review by kantai — COMMENTED** [review]
This looks mostly good to me, just had a question about signers who didn't validate a proposal.

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 6]
```diff
@@ -3198,14 +3189,14 @@ fn tx_replay_forking_test() {
...
     let post_fork_1_nonce = get_account(&http_origin, &sender_addr).nonce;
 
-    let burn_blocks = test_observer::get_burn_blocks().clone();
+    let burn_blocks = test_observer::get_burn_blocks();
 
     for block in burn_blocks {
-        let height = block.get("burn_block_height").unwrap().as_number().unwrap();
-        if height.as_u64().unwrap() < 230 {
+        let height = block.burn_block_height;
+        if height < 230 {
```
[the remark this answers is not in the document]
Comment (hstove) [reply testnet/stacks-node/src/tests/signer/v0.rs]: This was just for debugging - I've removed it now

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 7]
```diff
@@ -3362,6 +3335,235 @@ fn tx_replay_forking_test() {
...
+        let commits_count = submitted_commits.load(Ordering::SeqCst);
+        next_block_and_controller(
+            &signer_test.running_nodes.btc_regtest_controller,
+            60,
+            |_btc_controller| {
+                let commits_submitted = submitted_commits.load(Ordering::SeqCst);
+                Ok(commits_submitted > commits_count)
+            },
+        )
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (hstove) [reply testnet/stacks-node/src/tests/signer/v0.rs]: Good catch!

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 8]
```diff
@@ -3290,104 +3226,477 @@ fn tx_replay_forking_test() {
...
+        .expect("Timed out waiting for block pushed after fork");
+    assert!(!block.txs.iter().any(|tx| tx.txid().to_string() == txid));
+    assert!(!block.txs.iter().any(|tx| tx.txid().to_string() == txid_2));
+    info!(
+        "---- Wait for block proposal at stacks block height {} ----",
+        stacks_height_before + 2
+    );
+    // Next the signers will attempt to propose a block that does not contain the necessary replay tx and signers will reject it
+    let rejected_block = wait_for_block_proposal(30, stacks_height_before + 2, &stacks_miner_pk)
+        .expect("Timed out waiting for block pushed after fork");
```
Comment (fdefelici) [inline testnet/stacks-node/src/tests/signer/v0.rs:3466]: nit
```suggestion
        .expect("Timed out waiting for block proposal after fork");
```

File: stacks-signer/src/v0/signer_state.rs [hunk 9]
```diff
@@ -366,10 +385,45 @@ impl LocalStateMachine {
...
+                            tx.payload,
+                            TransactionPayload::TenureChange(TenureChangePayload {
+                                cause: TenureChangeCause::BlockFound,
+                                ..
+                            }) | TransactionPayload::Coinbase(..)
+                        )
+                    }) {
+                        info!("Signer state: clearing the tx replay set due to unrecognized transactions");
+                        prior_state_machine.tx_replay_set = ReplayTransactionSet::none();
+                    }
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs:417]: I wanted to highlight this change, related to [this comment](https://github.com/stacks-network/stacks-core/pull/6109#discussion_r2089449876) about "what happens if a block exhausts the replay set, but I didn't validate it". 

This also addresses "what happens if signers disagree about tx replay" - in which case, the signer's local state machine needs to know when to get out of tx replay.

Basically, this code is saying "if a block with "normal" transactions comes in, and we think we're in tx replay but this block didn't exhaust our replay set, we should capitulate and get out of tx replay".

File: testnet/stacks-node/src/tests/signer/v0.rs [hunk 10]
```diff
@@ -3290,104 +3226,474 @@ fn tx_replay_forking_test() {
...
+///   in the tx replay set
+/// - Force the miner to ignore replay transactions and attempt
+///   to mine a regular block
+/// - Verify the signers reject this proposed block due to it
+///   missing the replay transactions
+/// - Allow the miner to consider the replay transactions
+/// - Verify the miner correctly constructs a block containing the
+///   tx replay set
+/// - Verify the signers approve subsequent blocks
+fn tx_replay_e2e_test() {
```
Comment (fdefelici) [inline testnet/stacks-node/src/tests/signer/v0.rs:3331]: nit: The name `tx_replay_e2e_test` feels a bit too generic. Could we use a more descriptive name that reflects the specific scenario being tested? 

Could this suggestion better convey the intent of the test?
```suggestion
fn tx_replay_ongoing_rejects_regular_block_proposal() {
```

File: stacks-signer/src/v0/signer_state.rs (hunk 9 above)
Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs:417]: Ah, of course, good catch. This is not the intended behavior, and I'll have to re-think this code block.

Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs:417]: Eventually if it could help with the reasoning, this is a patched version of the `Ok(false)` branch that I'm using locally for my task #5971 

```
Ok(false) => {
    info!("Signer state: got a new block during replay that don't complete the replay set.";
        "txs" => ?txs,
    );
    
    // Closures
    let is_tenure_change_block_found = || {
        txs.len() == 2 &&
        matches!(
                txs[0].payload,
                TransactionPayload::TenureChange(TenureChangePayload {
                    cause: TenureChangeCause::BlockFound,
                    ..
                })) &&
        matches!(
                txs[1].payload,
                TransactionPayload::Coinbase(..))
    };

    let is_tenure_change_extend = || {
        txs.len() >= 1 &&
        matches!(
                txs[0].payload,
                TransactionPayload::TenureChange(TenureChangePayload {
                    cause: TenureChangeCause::Extended,
                    ..
                }))
    };

    let is_tx_replay_subset = || {
        let replay_set= prior_state_machine
            .tx_replay_set.as_slice();
            
        replay_set.windows(txs.len()).any(|w| w == txs)
    };

    if is_tenure_change_block_found() {
        info!("Signer state: got Tenure Change Block found. Allowed during Tx Replay.";
            "txs" => ?txs,
        );
    } else if is_tenure_change_extend() {
        info!("Signer state: got Tenure Change Extend. Allowed during Tx Replay.";
            "txs" => ?txs,
        );
        //TODO: Handle case where the tenure extend block, contains Tx Replay txs or not
        //In issue #5972
    } else if is_tx_replay_subset() {
        //if Txs is an ordered subset of Tx Replay set OK
        info!("Signer state: got Tx Replay subset. Allowed during Tx Replay.";
            "txs" => ?txs,
        );
        
        let replay_set= prior_state_machine
            .tx_replay_set.as_slice();
        let split_index = txs.len();
        prior_state_machine.tx_replay_set = 
            ReplayTransactionSet::new(replay_set[split_index..].to_vec());
    } else {
        //Otherwise capitulate, clearing tx replay set
        info!("Signer state: clearing the tx replay set due to unrecognized transactions");
        prior_state_machine.tx_replay_set = ReplayTransactionSet::none();
    }
}
```

File: stackslib/src/chainstate/stacks/miner.rs [hunk 11]
```diff
@@ -2311,7 +2315,12 @@ impl StacksBlockBuilder {
             }
         }
 
-        let result = if replay_transactions.is_empty() {
+        #[cfg(any(test, feature = "testing"))]
+        let use_mempool_txs = replay_transactions.is_empty() || TEST_EXCLUDE_REPLAY_TXS.get();
+        #[cfg(not(any(test, feature = "testing")))]
+        let use_mempool_txs = replay_transactions.is_empty();
```
[the remark this answers is not in the document]
Comment (hstove) [reply stackslib/src/chainstate/stacks/miner.rs:2321]: > Is this behavior intended?

Yes! And in fact, it's a pretty standard scenario, since the forked block's transactions end up back in the mempool.

If a miner is on an old version, and they end up constructing a valid replay block, it's better to let it happen (as opposed to a temporary chain stall)

File: stacks-signer/src/v0/signer_state.rs (hunk 4 above)
Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: Just to follow up with a written response - the updated logic in this PR is to have three separate branches:

- block validated with replay set, not exhausted: do nothing
- block validated with replay set, exhausted: clear replay set
- block not validated with our local replay set: clear replay set

So, to answer the question:

> What happens to signers who don't validate block A (maybe they were offline?), where block A is the block that exhausts the replay set.

When the signer processes block A, it'll hit that third branch, and their local state machine's replay set will be cleared.

**Comment by hstove** [comment]
I've pushed a new update based on some feedback from @fdefelici , and it would be great to get reviews on this! cc @kantai @jferrant

File: stackslib/src/chainstate/stacks/miner.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (fdefelici) [reply stackslib/src/chainstate/stacks/miner.rs:2321]: implemented integration test that cover this scenario in this #6167

File: libsigner/src/v0/signer_state.rs [hunk 12]
```diff
@@ -132,7 +132,12 @@ impl GlobalStateEvaluator {
                         burn_block_height,
                         current_miner,
                         ..
-                    } => (burn_block, burn_block_height, current_miner, None),
+                    } => (
+                        burn_block,
+                        burn_block_height,
+                        current_miner,
+                        ReplayTransactionSet::new(vec![]),
```
Comment (obycode) [inline libsigner/src/v0/signer_state.rs]: nit: `::none()`

File: libsigner/src/v0/signer_state.rs [hunk 13]
```diff
@@ -177,6 +182,85 @@ impl GlobalStateEvaluator {
...
+    /// Unwrap the `ReplayTransactionSet`. Panics if the set is empty.
+    pub fn unwrap(self) -> Vec<StacksTransaction> {
+        if self.is_empty() {
+            panic!("Called `unwrap` on an empty `ReplayTransactionSet`");
+        }
+        self.0
+    }
+
+    /// Map into an optional, returning `None` if the set is empty
+    pub fn into_optional(&self) -> Option<Vec<StacksTransaction>> {
```
Comment (obycode) [inline libsigner/src/v0/signer_state.rs]: I think this naming is a bit confusing, since `into` usually consumes the value. Maybe `as_optional` or even `clone_as_optional` would be more appropriate if we can't avoid this clone.

File: libsigner/src/v0/signer_state.rs [hunk 14]
```diff
@@ -177,6 +182,85 @@ impl GlobalStateEvaluator {
...
+
+    /// Check if the `ReplayTransactionSet` is empty
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Unwrap the `ReplayTransactionSet`. Panics if the set is empty.
+    pub fn unwrap(self) -> Vec<StacksTransaction> {
+        if self.is_empty() {
+            panic!("Called `unwrap` on an empty `ReplayTransactionSet`");
```
Comment (obycode) [inline libsigner/src/v0/signer_state.rs]: This worries me a bit. It looks like this is only called from `unwrap_or_default`, so could we instead just delete this method and move the code there?

File: stackslib/src/net/api/postblock_proposal.rs [hunk 15]
```diff
@@ -173,6 +175,12 @@ pub struct BlockValidateOk {
     pub cost: ExecutionCost,
     pub size: u64,
     pub validation_time_ms: u64,
+    /// If a block was validated by a transaction replay set,
+    /// then this return Some with the hash of the replay set.
```
Comment (obycode) [inline stackslib/src/net/api/postblock_proposal.rs]:
```suggestion
    /// then this returns Some with the hash of the replay set.
```

File: stacks-signer/src/v0/signer_state.rs [hunk 16]
```diff
@@ -625,7 +674,7 @@ impl LocalStateMachine {
                 burn_block_height: *burn_block_height,
                 current_miner: current_miner.into(),
                 active_signer_protocol_version,
-                tx_replay_set: tx_replay_set.cloned(),
+                tx_replay_set: tx_replay_set.clone(),
```
Comment (obycode) [inline stacks-signer/src/v0/signer_state.rs]: I don't think this clone is necessary.

File: stackslib/src/net/api/tests/postblock_proposal.rs [hunk 17]
```diff
@@ -957,3 +968,85 @@ fn replay_validation_test_budget_exceeded() {
         }
     }
 }
+
+#[test]
+#[ignore]
+/// Replay set has [deploy, big_a, big_b]
+/// The block has [deploy, big_a]
```
Comment (obycode) [inline stackslib/src/net/api/tests/postblock_proposal.rs:975]: Would it make sense to add another test that checks that the following block, with `[big_b]` is handled correctly as well?

Comment (hstove) [reply stackslib/src/net/api/tests/postblock_proposal.rs:975]: Yes it would, but I actually have had a pretty tough time figuring out how to do this with `TestRPC`. Regardless, this will be covered in tests for #5972

**Comment by hstove** [comment]
@obycode I've addressed all but your last comment - mind re-reviewing?

**Comment by jferrant** [comment]
I am assuming the tests are not newly flakey. (I see bitcoin_reorg_extended_tenure failing on another PR which is just a clippy fix so I know for sure that one is already flakey)

**Comment by obycode** [comment]
> I am assuming the tests are not newly flakey.

Or maybe everyone just knows how much you love fixing flaky tests so Hank left them for you.
